### ISSUE TITLE
Invalid stage identifier with BasePathMapping depending on SAM ApiResource

### ISSUE THREAD

**Issue by kennu** [awslabs/aws-cdk#931]
Creating an **apigateway.cloudformation.BasePathMappingResource** that depends on a **sam.cloudformation.ApiResource** generates this error:

    [ 6/11] Mon Oct 15 2018 04:40:45 GMT+0300 (EEST)  CREATE_FAILED       
    [AWS::ApiGateway::BasePathMapping] ApiBasePath3EEAADC4 Invalid stage identifier specified 
    (Service: AmazonApiGateway; Status Code: 400; Error Code: BadRequestException; Request ID: 
    55db9c2c-d01b-11e8-9ab5-c3df5de0ac4a)

I believe this problem is caused by this unfixed dependency issue in SAM: https://github.com/awslabs/serverless-application-model/issues/192

I have not yet found a workaround. Is there a way to manually specify a dependency to the stage resource that SAM generates?

Tested with AWS CDK 0.12

**Comment by kennu** [comment]
I have now converted my project to use **apigateway.RestApi** instead of **sam.cloudformation.ApiResource**. Thanks for the related assistance also in here @eladb : https://stackoverflow.com/questions/52752201/enabling-cors-for-aws-api-gateway-with-the-aws

**Comment by eladb** [comment]
@kennu glad to hear. Does the CORS solution work? Would love to add support for it as part of `RestApi`

**Comment by kennu** [comment]
Yes, CORS worked. It's nice that in AWS CDK I was able to add a little helper function like this:

    function addCorsOptions(apiResource: apigateway.IRestApiResource) {
      const options = apiResource.addMethod('OPTIONS', new apigateway.MockIntegration({
        integrationResponses: [{
          statusCode: '200',
          responseParameters: {
            'method.response.header.Access-Control-Allow-Headers': "'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token,X-Amz-User-Agent'",
            'method.response.header.Access-Control-Allow-Origin': "'*'",
            'method.response.header.Access-Control-Allow-Credentials': "'false'",
            'method.response.header.Access-Control-Allow-Methods': "'OPTIONS,GET,PUT,POST,DELETE'",
          },
        }],
        passthroughBehavior: apigateway.PassthroughBehavior.Never,
        requestTemplates: {
          "application/json": "{\"statusCode\": 200}"
        },
      }))
      const methodResource = options.findChild('Resource') as apigateway.cloudformation.MethodResource
      methodResource.propertyOverrides.methodResponses = [{
        statusCode: '200',
        responseModels: {
          'application/json': 'Empty'
        },
        responseParameters: {
          'method.response.header.Access-Control-Allow-Headers': true,
          'method.response.header.Access-Control-Allow-Methods': true,
          'method.response.header.Access-Control-Allow-Credentials': true,
          'method.response.header.Access-Control-Allow-Origin': true,
        },
      }]
    }

**Comment by kennu** [comment]
Btw, just in case someone googles this issue and still gets the the error message: If you use **apigateway.RestApi** instead of SAM, you need to make sure you configure the stageName properly in deployOptions:

    const api = new apigateway.RestApi(this, 'Api', {
      restApiName: props.stage + '-my-api',
      deployOptions: {
        stageName: props.stage,
      },
    })

The stage name must match what you put in the base path resource:

    new apigateway.cloudformation.BasePathMappingResource(this, 'BasePath', {
      basePath: 'my-basepath',
      domainName: 'mydomain.example.com',
      restApiId: api.restApiId,
      stage: props.stage,
    })

By default AWS CDK generates a stage called 'prod' unless you override it. I've used a property to configure it.
